[FX] Fix tracing support for ScriptMethods

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52884 [FX] Fix tracing support for ScriptMethods**

Previously, TorchBind method invocations actually weren't captured by symbolic tracing because TorchBind methods were implemented as `ScriptMethod`s. `ScriptMethods` did NOT participate in `__torch_function__` dispatch, thus those calls weren't actually recorded. The original `test_torchbind_class_attribute_in_fx` was a red herring, since the method called did not take any tensor-valued arguments (it took no arguments at all) and the test checked for correctness based on value. The value captured was actually just a baked-in constant.

This fixes all of that by adding `__torch_function__` dispatch support for `ScriptMethod`s. It also adds a test that calls a method that takes a `Tensor` to make sure this actually works

Differential Revision: [D26675801](https://our.internmc.facebook.com/intern/diff/D26675801)

Commit:
[WIP][FX] Fix tracing support for torchbind

[ghstack-poisoned]

Commit:
Update on "[WIP][FX] Fix tracing support for torchbind"


Differential Revision: [D26675801](https://our.internmc.facebook.com/intern/diff/D26675801)

[ghstack-poisoned]

Commit:
Update on "[WIP][FX] Fix tracing support for torchbind"


Differential Revision: [D26675801](https://our.internmc.facebook.com/intern/diff/D26675801)

[ghstack-poisoned]

Commit:
Update on "[WIP][FX] Fix tracing support for torchbind"


Differential Revision: [D26675801](https://our.internmc.facebook.com/intern/diff/D26675801)

[ghstack-poisoned]

Commit:
Update on "[WIP][FX] Fix tracing support for torchbind"


Previously, TorchBind method invocations actually weren't captured by symbolic tracing because TorchBind methods were implemented as `ScriptMethod`s. `ScriptMethods` did NOT participate in `__torch_function__` dispatch, thus those calls weren't actually recorded. The original `test_torchbind_class_attribute_in_fx` was a red herring, since the method called did not take any tensor-valued arguments (it took no arguments at all) and the test checked for correctness based on value. The value captured was actually just a baked-in constant.

This fixes all of that by adding `__torch_function__` dispatch support for `ScriptMethod`s. It also adds a test that calls a method that takes a `Tensor` to make sure this actually works

Differential Revision: [D26675801](https://our.internmc.facebook.com/intern/diff/D26675801)

[ghstack-poisoned]

jamesr66a commented:
@hameerabbasi @ezyang Can you take a look at this? I'm making it so that TorchScript callables (in this patch, method calls) are admissible to `__torch_function__` dispatch. This allows us to record TorchBind method invocations in FX, for one. I'm particularly interested in if there are deeper implications here wrt the design of `__torch_function__`. Please let me know!

Commit:
Update on "[WIP][FX] Fix tracing support for torchbind"


Previously, TorchBind method invocations actually weren't captured by symbolic tracing because TorchBind methods were implemented as `ScriptMethod`s. `ScriptMethods` did NOT participate in `__torch_function__` dispatch, thus those calls weren't actually recorded. The original `test_torchbind_class_attribute_in_fx` was a red herring, since the method called did not take any tensor-valued arguments (it took no arguments at all) and the test checked for correctness based on value. The value captured was actually just a baked-in constant.

This fixes all of that by adding `__torch_function__` dispatch support for `ScriptMethod`s. It also adds a test that calls a method that takes a `Tensor` to make sure this actually works

Differential Revision: [D26675801](https://our.internmc.facebook.com/intern/diff/D26675801)

[ghstack-poisoned]

Commit:
Update on "[WIP][FX] Fix tracing support for ScriptMethods"


Previously, TorchBind method invocations actually weren't captured by symbolic tracing because TorchBind methods were implemented as `ScriptMethod`s. `ScriptMethods` did NOT participate in `__torch_function__` dispatch, thus those calls weren't actually recorded. The original `test_torchbind_class_attribute_in_fx` was a red herring, since the method called did not take any tensor-valued arguments (it took no arguments at all) and the test checked for correctness based on value. The value captured was actually just a baked-in constant.

This fixes all of that by adding `__torch_function__` dispatch support for `ScriptMethod`s. It also adds a test that calls a method that takes a `Tensor` to make sure this actually works

Differential Revision: [D26675801](https://our.internmc.facebook.com/intern/diff/D26675801)

[ghstack-poisoned]

hameerabbasi commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -948,10 +950,71 @@ inline py::object runAndInsertCall(
+  std::vector<py::handle> overloaded_args;
+  size_t total_arg_num = args.size() + kwargs.size();
+  for (size_t i = 0; i < args.size(); ++i) {
+    is_tensor_and_append_overloaded(args[i].ptr(), &overloaded_args);
```
This will fail for `TensorList` operands.

hameerabbasi replied:
Can this take `TensorList`s? I believe the reason to skip in the other place was simply that they didn’t need it.

hameerabbasi replied:
I see from [this comment](https://github.com/pytorch/pytorch/pull/52884#issuecomment-788128434) that no `TensorList`s are needed, if so, you can resolve this, @jamesr66a.

jamesr66a replied:
@hameerabbasi  I don't really understand this comment thread.

1) Can you explain exactly what would fail about `TensorList`? I didn't write this code, but from what I can tell, `is_tensor_and_append_overloaded` and `is_tensor_list_and_append_overloaded` are both best-effort and both if `args` is a `Tensor` or if `arg` is a `TensorList` should be covered by the two cases here. Can you explain a bit more about what you're referring to?
2) I don't understand the implication of Ed's comment that suggests that `TensorList` operands are not possible.

hameerabbasi replied:
`is_tensor_and_append_overloaded` never checks for `TensorList`, only for something that "looks like a `Tensor`", i.e. has `__torch_function__` and isn't `nn.Parameter` or `Tensor`. See this comment: https://github.com/pytorch/pytorch/blob/a0a1bb074b954338d7f19dac9a9c8136376bce7c/torch/csrc/utils/python_arg_parser.h#L748-L757

If `arg` turns out to be a `TensorList`, then it won't recurse inside the `TensorList` and therefore miss types that have `__torch_function__`.

What I'm looking for here is confirmation that `arg` will never be a `TensorList`, i.e. that the callable that you're overriding here will never take a `TensorList` as a potential argument. Ed's comment has an example of what the callable you're overriding might look like, and if it only takes a `Tensor` in, then we're okay.

jamesr66a replied:
@hameerabbasi still don't understand your comment. In this code, there's a call to `is_tensor_list_and_append_overloaded` directly after this comment thread, which AFAICT does do this recursion.

hameerabbasi replied:
Ah, missed that one. I apologize for the confusion; consider this resolved.

hameerabbasi commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -948,10 +950,71 @@ inline py::object runAndInsertCall(
+  // match the operator in `operations` first, then the order will
+  // be guaranteed.
+  for (auto item : kwargs) {
+    is_tensor_and_append_overloaded(item.second.ptr(), &overloaded_args);
```
Same here, `TensorList` will fail. You may need to add an explicit check for `isinstance(arg, (tuple, list))` and recurse inside too.

hameerabbasi replied:
As well as this one.

hameerabbasi commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -948,10 +950,71 @@ inline py::object runAndInsertCall(
+  // Handle __torch_function__ dispatch
+  std::vector<py::handle> overloaded_args;
+  size_t total_arg_num = args.size() + kwargs.size();
+  for (size_t i = 0; i < args.size(); ++i) {
```
Perhaps switch to a range-based for over `args_vec`? It's both faster and more readable.

jamesr66a replied:
I don't know about faster (I would imagine the compiler would be able to do a trivial strength reduction here) but for style I'll change it

hameerabbasi commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -948,10 +950,71 @@ inline py::object runAndInsertCall(
+        false /* throw_error */);
+  }
+  if (overloaded_args.size() > 0) {
+    std::vector<py::object> overloaded_types;
+    overloaded_types.reserve(overloaded_args.size());
+    for (auto& oarg : overloaded_args) {
+      overloaded_types.push_back(
+          py::reinterpret_borrow<py::object>((PyObject*)Py_TYPE(oarg.ptr())));
+    }
```
Types need to be weakly ordered -- Subclasses before superclasses. That invariant isn't satisfied here, unfortunately.

hameerabbasi replied:
Scratch that -- Looks like you aren't using `overloaded_types` at all.

hameerabbasi replied:
`overloaded_types` can be removed, I believe.

hameerabbasi reviewed (changes requested):
This will require updates to the RFC -- Currently the only methods we support are on `Tensor`. We might need a small update to say `ScriptMethod`s are passed in as `func` since they're callable.

We also might need a testing script to test `ScriptMethod`s.

jamesr66a commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -948,10 +950,71 @@ inline py::object runAndInsertCall(
+            qualname.name().c_str(),
+            args[0].ptr(),
+            qualname.prefix().c_str()));
+  }
```
@hameerabbasi To be clear, this function is a direct copy-paste from the implementation for custom ops: https://github.com/pytorch/pytorch/blob/1d6bd157902d4b1347a5d03122d02b407658e263/torch/csrc/jit/python/init.cpp#L1020

So any deficiencies you're pointing out also apply there and should probably be split off into issues, as this points out deficiencies in the current implementation of `__torch_function__`.

hameerabbasi replied:
Code replication is usually bad -- We should think about extracting this somewhere so there aren't 4+ implementations to change when we fix these issues:

- Main C++ one
- This one
- The one you refer to in your comment
- The Python one
- Did I miss any? 🙂 

In any case, we can split it out into issues; but we should deduplicate first.

jamesr66a replied:
@hameerabbasi If it's okay, I'd prefer to land something like this PR before deduplicating. Lack of support for these methods is blocking some important use cases now. IMO landing duplicated code now should be OK to unblock so long as we make sure that someone goes in to deduplicate afterward, and make sure nobody's diverging the implementations in the intervening time. Does that sound good to you?

hameerabbasi replied:
It sounds good, I’ll open an issue for that tomorrow.

esqu1 replied:
@hameerabbasi What James said above; we actually really need this functionality internally for enabling some of our models to be FX+TorchScript compatible. It would be great if we could have this landed this soon and then revisit the refactoring in a later PR.

hameerabbasi replied:
I won't block this PR based on that, I'll comment here once the issue is open.

jamesr66a commented:
@hameerabbasi please let me know how to move forward on this. Is updating the RFC/making the final call an action item on your side?

ezyang commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -1090,6 +1090,22 @@ void initJitScriptBindings(PyObject* module) {
+      .def(
+          "__eq__",
+          [](const Module& self, const py::object& other) {
+            // TODO: call UDF if it exists
```
Is there a way to easily raise an error if you hit the TODO case? Maybe not...

jamesr66a replied:
I'm actually not sure if UDF `__eq__` is even a thing on `Module`. I could understand it for `Object` but not sure if it makes sense for `Module`

Speaking of which @gmagogsfm can we get a review on this from the TorchScript side? Thanks!

hameerabbasi commented:
I may have misunderstood -- Does "method" here mean in the Python sense (instance method) or is it a regular callable?

ezyang commented:
It's an honest to goodness method. What's going on here is kind of weird but basically it's like if you had:

```
class RELu:
  def run(t: Tensor):
    if has_torch_function_unary(t):
      return handle_torch_function(run, self, t)
    return t.relu()
```

I actually think this is pretty analogous to Tensor methods, because we don't really use bound methods in the API, I think.

hameerabbasi commented:
Then we need alterations to the RFC and test infrastructure. I believe there are some places  in the tests that assume `self` is a `Tensor`.

hameerabbasi commented:
I'd only like small changes here, considering [this comment](https://github.com/pytorch/pytorch/pull/52884#discussion_r585487701) is true. Specifically, moving to iterating over the vector and removing the types.

hameerabbasi commented:
For this to get my approval, I'd like:

- Iteration over `args_vec` instead of `args`: https://github.com/pytorch/pytorch/pull/52884#discussion_r584152093
- Removal of `overloaded_type` computation: https://github.com/pytorch/pytorch/pull/52884#discussion_r584975897
- Verification that the `TensorList` case will never be hit: https://github.com/pytorch/pytorch/pull/52884#discussion_r584151712 and https://github.com/pytorch/pytorch/pull/52884#discussion_r584151855

Aside from this, it's all good to go on my side.

Commit:
Update on "[FX] Fix tracing support for ScriptMethods"


Previously, TorchBind method invocations actually weren't captured by symbolic tracing because TorchBind methods were implemented as `ScriptMethod`s. `ScriptMethods` did NOT participate in `__torch_function__` dispatch, thus those calls weren't actually recorded. The original `test_torchbind_class_attribute_in_fx` was a red herring, since the method called did not take any tensor-valued arguments (it took no arguments at all) and the test checked for correctness based on value. The value captured was actually just a baked-in constant.

This fixes all of that by adding `__torch_function__` dispatch support for `ScriptMethod`s. It also adds a test that calls a method that takes a `Tensor` to make sure this actually works

Differential Revision: [D26675801](https://our.internmc.facebook.com/intern/diff/D26675801)

[ghstack-poisoned]

Commit:
Update on "[FX] Fix tracing support for ScriptMethods"


Previously, TorchBind method invocations actually weren't captured by symbolic tracing because TorchBind methods were implemented as `ScriptMethod`s. `ScriptMethods` did NOT participate in `__torch_function__` dispatch, thus those calls weren't actually recorded. The original `test_torchbind_class_attribute_in_fx` was a red herring, since the method called did not take any tensor-valued arguments (it took no arguments at all) and the test checked for correctness based on value. The value captured was actually just a baked-in constant.

This fixes all of that by adding `__torch_function__` dispatch support for `ScriptMethod`s. It also adds a test that calls a method that takes a `Tensor` to make sure this actually works

Differential Revision: [D26675801](https://our.internmc.facebook.com/intern/diff/D26675801)

[ghstack-poisoned]

SplitInfinity commented on test/test_fx.py:
```diff
@@ -1621,6 +1621,46 @@ def forward(self):
         self.checkGraphModule(m, ())
 
+    def test_torchbind_class_attribute_in_fx_tensor_arg(self):
+        if TEST_WITH_ROCM or IS_SANDCASTLE or IS_WINDOWS or IS_MACOS:
+            self.skipTest("torch.classes._TorchScriptTesting._ReLUClass is registered, skipping")
```
This will probably time out in fbcode (see: D26838029). I suggest using `@unittest.skipIf` instead.

SplitInfinity replied:
Actually I think the problem I'm referring to only happens in `test_jit.py`, you can probably ignore this.
